### PR TITLE
Copy generics from functions to Return Position Impl Traits in HIR lowering

### PR DESCRIPTION
**Pull request by spastorino**
Going to write some good description here. Opening to allow reviews as I write a proper description.

This is going to allow proper support for RPITITs and async fns in traits.
https://github.com/rust-lang/rust/pull/101224 can be rebased on top of this one.

r? @nikomatsakis 

cc @oli-obk @cjgillot @jackh726 @compiler-errors (unsure if somebody else is missing in the list of interested people)

I still need to make some improvements of code but please do not wait for that to review. I've already spotted some things that I want to fix.

### CODE REVIEW

**Comment by compiler-errors** [comment]
Is it possible to squash all of the "add tests" commits (and possibly others) into a few big ones? It's difficult to know which of these 47 commits make meaningful changes (i.e. worth while to review) and which can be skipped over. feel free to ignore this though, maybe others prefer everything to be split out.

**Comment by spastorino** [comment]
@compiler-errors I've tried to make commits easy to review one by one. As I was making step by step progress on things. It may not be perfect but tests were added in the right point in time with the code that added said functionality.

**Comment by compiler-errors** [comment]
Hm, then why not fold the test into the same commit that changes what the test is exercising? :thinking:

**Comment by spastorino** [comment]
@compiler-errors yeah, I've tried to do something like that, but I may have failed in some cases :). I'm willing to change this into something that's even easier to review but I don't think that that's squashing all the tests into one.

**Comment by spastorino** [comment]
Thanks for the try run @compiler-errors we need to do a crater run after try completes.

**Comment by compiler-errors** [comment]
@craterbot check

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 1]
```diff
@@ -626,9 +626,6 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
 
         let item = f(self);
         debug_assert_eq!(def_id, item.def_id());
-        // `f` should have consumed all the elements in these vectors when constructing `item`.
-        debug_assert!(self.impl_trait_defs.is_empty());
-        debug_assert!(self.impl_trait_bounds.is_empty());
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:568]: Why aren't those consumed any more?  This assertions is here to ensure we do not forget nodes during lowering.

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 2]
```diff
@@ -126,6 +134,8 @@ struct LoweringContext<'a, 'hir> {
     impl_trait_defs: Vec<hir::GenericParam<'hir>>,
     impl_trait_bounds: Vec<hir::WherePredicate<'hir>>,
 
+    in_scope_generics: Vec<LocalDefId>,
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:137]: Why do we need this?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 3]
```diff
@@ -810,28 +873,161 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
-        generic_params.extend(extra_lifetimes.into_iter().filter_map(|(ident, node_id, res)| {
-            self.lifetime_res_to_generic_param(ident, node_id, res)
-        }));
-        let generic_params = self.arena.alloc_from_iter(generic_params);
-        debug!(?generic_params);
 
-        generic_params
+        let mut hir_extra_lifetimes = Vec::new();
+
+        let current_node_id_to_def_id = self.resolver.node_id_to_def_id.clone();
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:995]: Why do we need this clone?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 4]
```diff
@@ -1434,13 +1709,53 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
             debug!(?collected_lifetimes);
             debug!(?new_remapping);
 
+            lctx.create_generic_defs(opaque_ty_def_id, &ast_generics.params, &mut new_remapping);
+            debug!(?new_remapping);
+
+            debug!("impl_trait_inputs = {:#?}", impl_trait_inputs);
+            for ty in impl_trait_inputs {
+                if let TyKind::ImplTrait(node_id, _) = &ty.kind {
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:1717]: Why not just store the `NodeId`s and bounds in `impl_trait_inputs`?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 5]
```diff
@@ -1461,20 +1776,93 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                    .chain(type_const_defs.into_iter())
+                    .chain(universal_params);
+
+                let mut predicates: Vec<_> = ast_generics
+                    .params
+                    .iter()
+                    .filter(|param| {
+                        matches!(
+                            param.kind,
+                            GenericParamKind::Const { .. } | GenericParamKind::Type { .. }
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:1835]: What about lifetime outlives predicates?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 6]
```diff
@@ -1502,11 +1891,46 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+            })
+            .collect();
         debug!(?lifetimes);
 
+        let type_const_args = self.collect_type_const_args(&ast_generics.params);
+
+        let universal_args: Vec<_> = impl_trait_inputs.iter().map(|ty| match ty.kind {
+            TyKind::ImplTrait(node_id, _) => {
+                let def_id = self.local_def_id(node_id);
+                let span = ty.span;
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:1903]:
```suggestion
                let span = self.lower_span(ty.span);
```
so you don't need to call it again, not `lower_ident` with the same span.

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 7]
```diff
@@ -1502,11 +1891,46 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                let ident = Ident::from_str_and_span(&pprust::ty_to_string(ty), span);
+                hir::GenericArg::Type(hir::Ty {
+                    hir_id: self.next_id(),
+                    span,
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:1907]: Related: this is currently missing a `lower_span`.

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 8]
```diff
@@ -1627,24 +2051,30 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+            ast_inputs = &ast_inputs[..ast_inputs.len() - 1];
         }
-        let inputs = self.arena.alloc_from_iter(inputs.iter().map(|param| {
+        let mut impl_trait_inputs = vec![];
+        let inputs = self.arena.alloc_from_iter(ast_inputs.iter().map(|param| {
             if fn_node_id.is_some() {
-                self.lower_ty_direct(&param.ty, ImplTraitContext::Universal)
+                self.lower_ty_direct(
+                    &param.ty,
+                    &mut ImplTraitContext::Universal { apit_nodes: &mut impl_trait_inputs },
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:2072]: Nit: could we create the `itctxt` outside of the iterator, for both branches?

File: compiler/rustc_resolve/src/late/lifetimes.rs [hunk 9]
```diff
@@ -1046,20 +1049,31 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
...
-                                        lifetime.name.ident(),
-                                    ))
-                                    .emit();
+
+                                let hir_map = this.tcx.hir();
+                                let parent = hir_map.get_parent_node(lt.hir_id);
+
+                                match hir_map.get(parent) {
+                                    hir::Node::Item(&hir::Item { kind: hir::ItemKind::OpaqueTy(..), .. }) => {
+                                        // do not warn
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late/lifetimes.rs:1058]: Why do we need this?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 10]
```diff
@@ -810,28 +873,161 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                }
+            })
+            .collect()
+    }
+
     /// Lowers a lifetime binder that defines `generic_params`, returning the corresponding HIR
     /// nodes. The returned list includes any "extra" lifetime parameters that were added by the
     /// name resolver owing to lifetime elision; this also populates the resolver's node-id->def-id
     /// map, so that later calls to `opt_node_id_to_def_id` that refer to these extra lifetime
     /// parameters will be successful.
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:978]: Could you complete the comment here?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 11]
```diff
@@ -1434,13 +1709,53 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                    new_remapping.insert(old_def_id, new_def_id);
+                } else {
+                    unreachable!(
+                        "impl_trait_inputs contains {:?} which is not TyKind::ImplTrait(..)",
+                        ty.kind
+                    );
+                }
+            }
+
+            let in_scope_generics: Vec<_> = ast_generics
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:1734]: Why do we need this?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 12]
```diff
@@ -254,14 +271,18 @@ impl ResolverAstLoweringExt for ResolverAstLowering {
...
     /// Example: `fn foo(x: impl Debug)`, where `impl Debug` is conceptually
     /// equivalent to a fresh universal parameter like `fn foo<T: Debug>(x: T)`.
     ///
     /// Newly generated parameters should be inserted into the given `Vec`.
-    Universal,
+    Universal {
+        apit_nodes: &'b mut Vec<&'itctx Ty>,
+    },
+
+    UniversalInRPIT,
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:285]: What is the difference with `Universal`?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 13]
```diff
@@ -254,14 +271,18 @@ impl ResolverAstLoweringExt for ResolverAstLowering {
...
+#[derive(Debug)]
+enum ImplTraitContext<'b, 'itctx> {
     /// Treat `impl Trait` as shorthand for a new universal generic parameter.
     /// Example: `fn foo(x: impl Debug)`, where `impl Debug` is conceptually
     /// equivalent to a fresh universal parameter like `fn foo<T: Debug>(x: T)`.
     ///
     /// Newly generated parameters should be inserted into the given `Vec`.
-    Universal,
+    Universal {
+        apit_nodes: &'b mut Vec<&'itctx Ty>,
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:282]: Can't this be put in `LoweringContext`, with `impl_trait_defs` and `impl_trait_bounds` since they are related?

**Review by cjgillot — COMMENTED** [review]
This PR is huge, and will probably take a while to review. Do you have a possibility to split it up?

I wonder if all this business around RPIT/TAIT/async should go in a separate module, just to isolate it from the rest of lowering?

Do we have a test with `fn foo<T: Trait<fn(&u8) -> &u8>>()`? I don't understand how this case works.

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 2 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:137]: This is used to keep track of in scope lifetimes and then filter the ones that we've already created defs for (the ones that show up in hir bounds) to lower those lifetimes as `static` https://github.com/rust-lang/rust/pull/101345/files#diff-ad0c15bbde97a607d4758ec7eaf88248be5d6b8ae084dfc84127f81e3f7a9bb4R2472. I guess you already know about this thing that is being done, this thing exist in master and we would need at some point to remove this hack.

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 11 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:1734]: This was answered in https://github.com/rust-lang/rust/pull/101345#discussion_r963098364 or are you referring to why is it constructed here?.

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 4 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:1717]: Yes, I had a fixme to do that at some point. Going to make this change.

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 1 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs]: Right, this shouldn't have been removed. It was part of a hacky way I've used to fix a problem and then after doing the right thing, I should have introduced it back. Fixed https://github.com/rust-lang/rust/pull/101345/commits/0fadeede0b5631fd13eb78fcd38e7ae1608e5f44

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 3 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:998]: This was done in order to allow us to lower `for<'a>` twice in two different items. So the second lower can insert in the remapping.

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 7 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:1910]: https://github.com/rust-lang/rust/pull/101345/commits/835d7de9950804fa806aaf3ff6173b3299e99c43

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 8 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs]: Done https://github.com/rust-lang/rust/pull/101345/commits/971641f5307c81f4eac7d48c252b783a23a83ba4

File: compiler/rustc_resolve/src/late/lifetimes.rs (hunk 9 above)
Comment (spastorino) [reply compiler/rustc_resolve/src/late/lifetimes.rs:1058]: Because due to this `'static` lifetime trick we end with `'static` in the where clauses of the RPIT. This silences the warning for this case.

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 10 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:981]: Done https://github.com/rust-lang/rust/pull/101345/commits/d45f1d3a83aedff4b4cca2611e8736d2dd916864

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 12 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:285]: The difference is that `Universal` is used when lowering APITs and it carries `apit_nodes` vector which is filled as APITs are being lowered. `UniversalInRPIT` is used when lowering universal inside RPITs.
This is not great to be honest, we should probably want to do the same I've done for lifetimes, adding visitors and gather these apit_nodes on a side track.
I guess it would be better if we leave this for a followup PR.

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 13 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:282]: We (me and @nikomatsakis) were trying to do the opposite, actually. Trying to have more local data instead of growing `LoweringContext` more and more, which already is like a god object.
Anyway, as I've commented elsewhere, I think we should remove `apit_nodes` in favor of having a visitor that gathers this information. If you're ok, I'd leave for a followup.

**Comment by spastorino** [comment]
> This PR is huge, and will probably take a while to review. Do you have a possibility to split it up?

I'd be all to simplify this or split into different things. I don't see how given that this is all about this RPIT change. There are some commits that are unrelated but really minor and I don't think it'd simplify things. I'm all ears to make whatever changes are needed to simplify the review of this.

> I wonder if all this business around RPIT/TAIT/async should go in a separate module, just to isolate it from the rest of lowering?

Maybe, I'm willing to keep doing work on lowering, should we even meet and discuss some changes to simplify this code?. I was talking with you, @nikomatsakis and @oli-obk but maybe if we meet and make some simple plan I could work on things. 

> Do we have a test with `fn foo<T: Trait<fn(&u8) -> &u8>>()`? I don't understand how this case works.

I'm not sure I understand the exact case you want to test. Can you write the complete test case?. I'd assume that you want to return an RPIT, otherwise this PR would be completely unrelated but just in case if you can write the test I can properly check it.

BTW: I've gone over all your comments, commented and marked as resolved as a way to organize myself over the review better. Feel free to reopen unaddressed things or if you have more thoughts comment and unresolve.

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 14]
```diff
@@ -1502,11 +1896,46 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                .into_iter()
+                .chain(type_const_args.into_iter())
+                .chain(universal_args.into_iter()),
+        );
+
         // `impl Trait` now just becomes `Foo<'a, 'b, ..>`.
-        hir::TyKind::OpaqueDef(hir::ItemId { def_id: opaque_ty_def_id }, lifetimes)
+        let result = hir::TyKind::OpaqueDef(hir::ItemId { def_id: opaque_ty_def_id }, generic_args);
+        debug!("lower_opaque_impl_trait = {:?}", result);
+        result
```
Comment (oli-obk) [inline compiler/rustc_ast_lowering/src/lib.rs:1938]: you can add `ret` to the `#[instrument]` on the function to generate this very code that you wrote by hand here

File: compiler/rustc_middle/src/ty/generics.rs [hunk 15]
```diff
@@ -337,6 +337,9 @@ impl<'tcx> GenericPredicates<'tcx> {
         if let Some(def_id) = self.parent {
             tcx.predicates_of(def_id).instantiate_into(tcx, instantiated, substs);
         }
+
+        debug!("instantiate_into: predicates={:#?} substs={:#?}", instantiated.predicates, substs);
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/generics.rs:341]: This happens a lot, use `trace!`

File: compiler/rustc_resolve/src/late/lifetimes.rs [hunk 16]
```diff
@@ -525,6 +525,7 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
         }
     }
 
+    #[tracing::instrument(level = "debug", skip(self))]
```
Comment (oli-obk) [inline compiler/rustc_resolve/src/late/lifetimes.rs:528]: after a rebase you can just use `instrument` without the `tracing::` prefix

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 17]
```diff
@@ -2024,9 +2471,12 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
                 let p_name = ParamName::Plain(ident);
-                let param = self.resolver.get_remapped_def_id(param);
-
-                hir::LifetimeName::Param(param, p_name)
+                if self.in_scope_generics.contains(&param) {
+                    hir::LifetimeName::Static
+                } else {
+                    let param = self.resolver.get_remapped_def_id(param);
+                    hir::LifetimeName::Param(param, p_name)
+                }
```
Comment (oli-obk) [inline compiler/rustc_ast_lowering/src/lib.rs:2479]: Please document this. From the naming `in_scope_generics` it is not obvious why these are remapped to `'static`

File: compiler/rustc_resolve/src/late/lifetimes.rs (hunk 9 above)
Comment (oli-obk) [reply compiler/rustc_resolve/src/late/lifetimes.rs:1058]: Document this in the source

File: compiler/rustc_trait_selection/src/traits/wf.rs [hunk 18]
```diff
@@ -628,9 +631,29 @@ impl<'tcx> WfPredicates<'tcx> {
...
+                        | None => {
+                            // for opaque types introduced via `-> impl Trait` or async fn desugaring,
+                            // the where-clauses are currently kinda bogus so we skip them.
+                            //
+                            // These opaque type references only appear in 1 place (the fn return type)
+                            // and we know they are well-formed in that location if the fn's where-clauses are met.
+                            //
+                            // The only other place this type will propagate to is the return type of a fn call,
+                            // and it of course must satisfy the fn where clauses, so we know the type will be well-formed
+                            // in that context as well.\
```
Comment (oli-obk) [inline compiler/rustc_trait_selection/src/traits/wf.rs:655]: wait... if we are entirely ignoring where bounds, why are we copying them at all?

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 19]
```diff
@@ -2362,7 +2362,11 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         match path.res {
             Res::Def(DefKind::OpaqueTy, did) => {
                 // Check for desugared `impl Trait`.
-                assert!(ty::is_impl_trait_defn(tcx, did).is_none());
+                assert!(matches!(
+                    ty::impl_trait_origin(tcx, did),
+                    None | Some(hir::OpaqueTyOrigin::FnReturn(_))
+                        | Some(hir::OpaqueTyOrigin::TyAlias)
```
Comment (oli-obk) [inline compiler/rustc_typeck/src/astconv/mod.rs:2368]: why did the TyAlias case change?

Is this reachable for opaque types from other crates or can we avoid checking the `None` case now?

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 20]
```diff
@@ -2697,23 +2701,56 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                            bug!(
+                                "Generic argument {:?} should have been inferred already",
+                                generic_arg
+                            )
+                        }
+                    })
+                    .collect();
+                let ty = tcx.mk_opaque(def_id, tcx.intern_substs(&substs));
+                debug!(?ty);
+                return ty;
```
Comment (oli-obk) [inline compiler/rustc_typeck/src/astconv/mod.rs:2741]: again, use `ret` on `#[instrument]`

File: compiler/rustc_typeck/src/collect.rs [hunk 21]
```diff
@@ -2224,6 +2219,8 @@ fn gather_explicit_predicates_of(tcx: TyCtxt<'_>, def_id: DefId) -> ty::GenericP
         + has_own_self as u32
         + early_bound_lifetimes_from_generics(tcx, ast_generics).count() as u32;
 
+    debug!("gather_explicit_predicates_of(predicates={:?})", predicates);
+    debug!("gather_explicit_predicates_of(ast_generics={:?})", ast_generics);
```
Comment (oli-obk) [inline compiler/rustc_typeck/src/collect.rs:2223]: if you're using `instrument`, don't mention the function name in the tracing statements again
```suggestion
    debug!(?predicates);
    debug!(?ast_generics);
```

File: compiler/rustc_typeck/src/collect/item_bounds.rs [hunk 22]
```diff
@@ -65,9 +66,14 @@ fn opaque_type_bounds<'tcx>(
...
         let mut bounds = <dyn AstConv<'_>>::compute_bounds(&icx, item_ty, ast_bounds);
+
+        debug!("opaque_type_bounds(bounds={:?})", bounds);
         // Opaque types are implicitly sized unless a `?Sized` bound is found
         <dyn AstConv<'_>>::add_implicitly_sized(&icx, &mut bounds, ast_bounds, None, span);
-        tcx.arena.alloc_from_iter(bounds.predicates(tcx, item_ty))
+        debug!("opaque_type_bounds(bounds={:?})", bounds);
+        let result = tcx.arena.alloc_from_iter(bounds.predicates(tcx, item_ty));
+        debug!("opaque_type_bounds(predicates={:?})", result);
+        result
```
Comment (oli-obk) [inline compiler/rustc_typeck/src/collect/item_bounds.rs:76]: `#[instrument(ret)]`

**Review by oli-obk — COMMENTED** [review]
RIP (review in progress) 😆

**Comment by cjgillot** [comment]
@spastorino could you summarize the motivation for this PR?  It's a spread across several zulip threads over 6 months, and I'm afraid I don't have a comprehensive view any more, and the articulation you expect with https://github.com/rust-lang/rust/pull/101224.
